datafusion/physical-optimizer/src/progressive_evaluation/extract_ranges.rs

```diff
@@ -0,0 +1,900 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Extract [`NonOverlappingOrderedLexicalRanges`] from different sources.
+
+use arrow::compute::SortOptions;
+use arrow::datatypes::Schema;
+use datafusion_common::tree_node::TreeNode;
+use datafusion_common::{ColumnStatistics, Result, ScalarValue};
+use datafusion_datasource::PartitionedFile;
+use datafusion_physical_expr::expressions::Column;
+use datafusion_physical_expr::LexOrdering;
+use datafusion_physical_plan::{ExecutionPlan, ExecutionPlanProperties};
+use std::sync::Arc;
+
+use super::lexical_ranges::{LexicalRange, NonOverlappingOrderedLexicalRanges};
+use super::statistics::{column_statistics_min_max, PartitionStatisticsVisitor};
+
+/// Attempt to extract LexicalRanges for the given sort keys and input plan
+///
+/// Output will have N ranges where N is the number of output partitions
+///
+/// Returns None if not possible to determine ranges
+pub fn extract_disjoint_ranges_from_plan(
+    exprs: &LexOrdering,
+    input_plan: &Arc<dyn ExecutionPlan>,
+) -> Result<Option<NonOverlappingOrderedLexicalRanges>> {
```

Comment:
In this function, the lexical ranges (per sort key) are extracted as min/maxes without any other sense of how we will order them. (Just that we need the min/max value per partiiton).

It's then the `NonOverlappingOrderedLexicalRanges` which has a sense of proper ordering.